### ISSUE TITLE
Unique symbol vs Weak symbol

### ISSUE THREAD

**Issue by FFFrog** [pytorch/pytorch#125465]
## TL;DR

Unlike `clang++`, the `g++` compiler supports `STB_GNU_UNIQUE` by default, which may cause unexpected problems in some cases.

## Issue description

The test case named `test_open_device_serialization `in the [test_cpp_extensions_open_device_registration.py](https://github.com/pytorch/pytorch/blob/main/test/test_cpp_extensions_open_device_registration.py) will always fail if compiled by clang++, not with g++. The direct reason is that the variable `BackendMetaSerialization` in the `GetBackendMetaSerialization` function has different addresses in torch and other third-party modules.

## Prerequisites

Under Unix systems, Python imports a dynamic library module, which is actually implemented through the dlopen api, and there are huge differences in the different flags of dlopen, so we first need to obtain the default dlopen flag when importing in python.

Through the cpython code, we can learn that the default dlopen flag can be obtained through the following API.

```Python
>>> import sys
>>> sys.getdlopenflags()
2
```

By browsing the code of glibc, we can easily get what flag the number 2 represents (RTLD_NOW & RTLD_LOCAL):
https://github.com/lattera/glibc/blob/895ef79e04a953cac1493863bcae29ad85657ee1/bits/dlfcn.h#L24-L28

```C
/* The MODE argument to `dlopen' contains one of the following: */
#define RTLD_LAZY	0x00001	/* Lazy function call binding.  */
#define RTLD_NOW	0x00002	/* Immediate function call binding.  */
#define	RTLD_BINDING_MASK   0x3	/* Mask of binding time value.  */
#define RTLD_NOLOAD	0x00004	/* Do not load the object.  */
#define RTLD_DEEPBIND	0x00008	/* Use deep binding.  */
```

## Code example

We can use the following code to simulate the above test case to make it easier to locate this problem:

The directory structure is as follows:

```
code
├── 1.cc
├── demo.h
├── 2.cc
└── main.cc
```

**demo.h**
```C++
#include <iostream>

inline int& getValue()
{
    static int a = 1;
    printf("%p\n", &a);
    return a;
}
```

**1.cc**
```C++
#include "demo.h"

void add1()
{
    int& a = getValue();
    a += 10;
    std::cout << a << std::endl;
}
```

**2.cc**
```C++
#include "demo.h"

void add2()
{
    int& a = getValue();
    a += 20;
    std::cout << a << std::endl;
}
```

**main.cc**
```C++
#include <iostream>
#include <stdio.h>
#include <dlfcn.h>

extern void add1();
extern void add2();

int main() {
    void* h1;
    void* h2;

    typedef void (*func)();
    char* error;

    // ignore the error logic
    h1 = dlopen("./lib1.so", RTLD_NOW | RTLD_LOCAL);
    func add1 = (func)dlsym(h1, "_Z4add1v");

    h2 = dlopen("./lib2.so", RTLD_NOW | RTLD_LOCAL);
    func add2 = (func)dlsym(h2, "_Z4add2v");

    add1();
    add2();

    dlclose(h1);
    dlclose(h2);

    return 0;
}
```

## Code Analysis

We can get different results by compiling with clang and gcc respectively.

**gcc:**

Command & Result:

```Shell
export CC=g++ && $CC -shared -fpic code/1.cc -o lib1.so && $CC -shared -fpic code/2.cc -o lib2.so && $CC -o main code/main.cc && ./main

0x7f5e0aa06050
11
0x7f5e0aa06050
31
```

The Symbols are as follow:

```Shell
$ readelf -W -s lib1.so | grep _ZZ8getValuevE1a
    13: 0000000000004050     4 OBJECT  UNIQUE DEFAULT   24 _ZZ8getValuevE1a
    30: 0000000000004050     4 OBJECT  UNIQUE DEFAULT   24 _ZZ8getValuevE1a
```

**clang++:**

Command & Result:

```Shell
export CC=clang++ && $CC -shared -fpic code/1.cc -o lib1.so && $CC -shared -fpic code/2.cc -o lib2.so && $CC -o main code/main.cc && ./main

0x7f257fdee050
11
0x7f257fde9050
21
```

The Symbols are as follow:

```Shell
$ readelf -W -s lib1.so | grep _ZZ8getValuevE1a
    13: 0000000000004050     4 OBJECT  WEAK   DEFAULT   22 _ZZ8getValuevE1a
    30: 0000000000004050     4 OBJECT  WEAK   DEFAULT   22 _ZZ8getValuevE1a
```

For g++, the presence of **UNIQUE** symbols in lib1.so will cause dlopen to force the use of the **RTLD_GLOBAL** flag to load the library, and then add its symbols to the **Gloabl symbol space**; when subsequently loading lib2.so, it will first search for **_ZZ8getValuevE1a** in the **Global symbol space** and do the symbol address relocation, resulting in **_ZZ8getValuevE1a** in different modules having the same address;

For clang++, since there are no **UNIQUE** symbols in lib1.so, it will be loaded according to the flag passed by dlopen (**RTLD_LOCAL & RTLD_NOW**). Its symbols will only be added to the **Local symbol space** and are only visible to **main**; when lib2.so is subsequently loaded, although it will still be searched in the **Global symbol space** first, since there is no corresponding symbol in the **Gloabl space**, the symbol in lib2.so will be selected, resulting in **_ZZ8getValuevE1a** in different modules having different addresses.

## Possible solutions:

- Set the default dlopen flag by setting **sys.setdlopenflags** api
- Avoid defining static variables in inline functions, which will lead to multiple copies of static variables in multiple dynamic libraries (move the definition to .cc)

If you have any other approaches, please show me, thank you!

cc @malfet @seemethere @zou3519 @mruberry @mikaylagawarecki @msaroufim
